duckduckgo/macos-browser
# Add Fastlane and code freeze lane

Task/Issue URL: https://app.asana.com/0/0/1203448085506687/f
Tech Design URL:
CC:

**Description**:
This PR adds `fastlane` and a first action that run the `git` part of `Prepare Build` section in the release template. 
In details, the `code_freeze` action does: 
- Updates `develop`.
- Updates `submodules`.
- Creates the release branch.
- Updates embedded files. 
- Updates XCode metadata. 
- Commits and pushes. 



**Steps to test this PR**:
_1. Set up_
These steps ensure that testing this PR doesn't interfere with any actual release work. 
1.a Checkout a new test branch out of this one. 
1.b The lane checks the local git repository is on `develop` before creating the release branch and fails otherwise. In order to test, you have to update `DEFAULT_BRANCH` with the name of your new branch (the one created above). 
1.c The lane creates the release branch using the value of `RELEASE_BRANCH` as a prefix. You may want to update `RELEASE_BRANCH` to `test-release` or any other string to make sure that the test doesn't affect or collide with the actual release branch. 
1.d Commit the changes and push to origin (the lane expects a clean default branch that can be pulled). 

_2. Installation_
2.a `rbenv`: To make sure the execution environment is consistent, Ruby version is pinned in `ruby-version`.  
- Install `rbenv` via `brew`.
- Run `rbenv install` to set up the local environment. 
- Run `rbenv local` and verify that Ruby 3.0.4 is used. 
(Note: `rvm` should work as well if you prefer it). 
2.b `bundle`: Install `bundle` via `brew`.
2.c Run `bundle install` to install `fastlane`.

_3. Test_
3.a Run `bundle exec fastlane code_freeze`.
3.b Verify that the new version is evaluated correctly (default is bumping the patch number). 
3.c Say `No` when you are prompted to confirm the version number and verify that the execution is aborted. 
3.d Run `bundle exec fastlane code_freeze version:<whatever version you want>`
3.e Verify that the custom version overrides the default one. 
3.f Confirm the version number. 
3.c Verify that all the steps are executed correctly and that the release branch is created and pushed. 

_4. Clean up_
4.a Delete the test release branch from GitHub. 
4.b Delete the test branch from GitHub. 
4.c Delete the test release branch from your local repository. 
4.d Delete the test branch from your local repository.


---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)
**When ready for review, remember to post the PR in MM**
